File: go/day_8/main.go
package main

import (
	"log"
	"math/bits"
	"os"
	"strings"
)

var nums = map[uint8]int{
	0b1110111: 0,
	0b0100100: 1,
	0b1011101: 2,
	0b1101101: 3,
	0b0101110: 4,
	0b1101011: 5,
	0b1111011: 6,
	0b0100101: 7,
	0b1111111: 8,
	0b1101111: 9,
}

type display struct {
	digits [10]uint8
	value  [4]uint8
}

func main() {
	data := readData()

	log.Println("Part 1: ", part1(data))
	log.Println("Part 2: ", part2(data))
}

func readData() []display {
	contents, err := os.ReadFile("../../data/day_8.txt")
	if err != nil {
		panic(err)
	}

	return parseContents(string(contents))
}

func parseContents(contents string) []display {
	lines := strings.Split(strings.TrimSpace(contents), "\n")
	displays := make([]display, 0, len(lines))

	for _, line := range lines {
		lineParts := strings.Split(line, " | ")

		digits := [10]uint8{}
		value := [4]uint8{}

		for idx, digit := range strings.Split(lineParts[0], " ") {
			num := uint8(0)
			for _, chr := range digit {
				num |= (1 << (chr - 'a'))
			}
			digits[idx] = num
		}

		for idx, digit := range strings.Split(lineParts[1], " ") {
			num := uint8(0)
			for _, chr := range digit {
				num |= (1 << (chr - 'a'))
			}
			value[idx] = num
		}

		displays = append(displays, display{digits: digits, value: value})
	}

	return displays
}

func part1(input []display) int {
	total := 0
	for _, display := range input {
		for _, v := range display.value {
			switch bits.OnesCount8(v) {
			case 2, 3, 4, 7:
				total++
			}
		}
	}
	return total
}

func part2(input []display) int {
	total := 0
	for _, display := range input {
		one := find(display.digits[:], func(u uint8) bool { return bits.OnesCount8(u) == 2 })
		seven := find(display.digits[:], func(u uint8) bool { return bits.OnesCount8(u) == 3 })
		four := find(display.digits[:], func(u uint8) bool { return bits.OnesCount8(u) == 4 })

		// &^: bitwise difference
		// 1: __c__f_
		// 7: a_c__f_

		// a = 7 - 1
		a := seven &^ one

		// 5s:
		// 2: a_cde_g
		// 3: a_cd_fg
		// 5: ab_d_fg
		lenFives := findAll(display.digits[:], func(u uint8) bool { return bits.OnesCount8(u) == 5 })

		// 3 contains 1 => 3 identified
		three := find(lenFives, func(u uint8) bool { return u&one == one })

		// 3: a_cd_fg
		// 4: _bcd_f_
		// (4 n 3): __cd_f_

		// b = 4 - (4 n 3)
		b := four &^ (four & three)

		// g = 3 - (4 n 3) - a
		g := three &^ (four & three) &^ a

		// 5 contains b => 5 and 2 identified
		five := find(lenFives, func(u uint8) bool { return u&b > 0 })
		two := find(lenFives, func(u uint8) bool { return u != five && u != three })

		// (3 n 5): a__d_fg
		// (2 n 3): a_cd__g

		// c = 3 - (3 n 5)
		c := three &^ (three & five)

		// e = 2 - (2 n 3)
		e := two &^ (two & three)

		// f = 3 - (2 n 3)
		f := three &^ (two & three)

		// d = remainder
		d := uint8(0b1111111) &^ a &^ b &^ c &^ e &^ f &^ g

		mapping := map[uint8]uint8{
			a: 0b0000001,
			b: 0b0000010,
			c: 0b0000100,
			d: 0b0001000,
			e: 0b0010000,
			f: 0b0100000,
			g: 0b1000000,
		}

		for idx, v := range display.value {
			asBits := uint8(0)
			for k, b := range mapping {
				if v&k > 0 {
					asBits |= b
				}
			}
			n := nums[asBits]
			for i := 0; i < (4 - idx - 1); i++ {
				n *= 10
			}
			total += n
		}
	}
	return total
}

func find(in []uint8, filter func(uint8) bool) uint8 {
	for _, v := range in {
		if filter(v) {
			return v
		}
	}
	panic("no match")
}

func findAll(in []uint8, filter func(uint8) bool) []uint8 {
	out := make([]uint8, 0, len(in))
	for _, v := range in {
		if filter(v) {
			out = append(out, v)
		}
	}
	return out
}

File: ts/day_8/main.ts
const readData = (): Display[] =>
  parseContents(Deno.readTextFileSync("./data/day_8.txt").trim());

const parseContents = (contents: string): Display[] =>
  contents.split("\n").map((line) => new Display().fromString(line)).reduce(
    (acc, d) => {
      acc.push(d);
      return acc;
    },
    [] as Display[],
  );

const part1 = (data: Display[]): number =>
  data.map((display) =>
    display.value.filter((v) =>
      v.length == 2 || v.length == 3 || v.length == 4 || v.length == 7
    )
      .length
  ).reduce((acc, cur) => acc + cur);

const part2 = (data: Display[]): number =>
  data.map((display) => display.decode()).reduce((acc, cur) => acc + cur);

function main() {
  const data = readData();

  console.log("Part 1: ", part1(data));
  console.log("Part 2: ", part2(data));
}

const NUMS: Record<string, number> = {
  "abcefg": 0,
  "cf": 1,
  "acdeg": 2,
  "acdfg": 3,
  "bcdf": 4,
  "abdfg": 5,
  "abdefg": 6,
  "acf": 7,
  "abcdefg": 8,
  "abcdfg": 9,
};

type digits = [
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
];

type values = [string, string, string, string];

class Display {
  digits: digits;
  value: values;

  constructor() {
    this.digits = ["", "", "", "", "", "", "", "", "", ""];
    this.value = ["", "", "", ""];
  }

  fromString(line: string): Display {
    const [digits, values] = line.split(" | ");
    digits.split(" ").forEach((digit, idx) => this.digits[idx] = digit);
    values.split(" ").forEach((value, idx) => this.value[idx] = value);
    return this;
  }

  fromComponents(digits: digits, value: values): Display {
    this.digits = digits;
    this.value = value;
    return this;
  }

  decode(): number {
    const mapping: Map<string, string> = new Map();

    const one = this.digits.find((digit) => digit.length == 2)?.split(
      "",
    ) as string[];
    const seven = this.digits.find((digit) => digit.length == 3)?.split(
      "",
    ) as string[];
    const four = this.digits.find((digit) => digit.length == 4)?.split(
      "",
    ) as string[];

    const a = seven.find((char) => !one.includes(char)) as string;
    mapping.set(a, "a");

    const lenFives = this.digits.filter((digit) => digit.length == 5).map((
      digit,
    ) => digit.split(""));

    const three = lenFives.find((digit) =>
      one.every((char) => digit.includes(char))
    ) as string[];

    const fourAndThree = four.filter((char) => three.includes(char));

    const b = four.find((char) => !fourAndThree.includes(char)) as string;
    mapping.set(b, "b");

    const g = three.find((char) =>
      !fourAndThree.includes(char) && char != a
    ) as string;
    mapping.set(g, "g");

    const five = lenFives.find((digit) => digit.includes(b)) as string[];
    const two = lenFives.find((digit) =>
      digit != five && digit != three
    ) as string[];

    const threeAndFive = three.filter((char) => five.includes(char));

    const c = three.find((char) => !threeAndFive.includes(char)) as string;
    mapping.set(c, "c");

    const twoAndThree = three.filter((char) => two.includes(char));

    const e = two.find((char) => !twoAndThree.includes(char)) as string;
    mapping.set(e, "e");

    const f = three.find((char) => !twoAndThree.includes(char)) as string;
    mapping.set(f, "f");

    const d = ["a", "b", "c", "d", "e", "f", "g"].find((char) =>
      !mapping.has(char)
    ) as string;
    mapping.set(d, "d");

    return this.value.map((v) => {
      return NUMS[v.split("").map((char) => mapping.get(char)).sort().join("")];
    }).reduce((acc, cur) => acc * 10 + cur);
  }
}

export { main, parseContents, part1, part2, readData };
export { Display };
